test/server/listener_manager_impl_test.cc

```diff
@@ -178,12 +187,24 @@ class ListenerManagerImplWithRealFiltersTest : public ListenerManagerImplTest {
       EXPECT_CALL(*socket_, requestedApplicationProtocols()).Times(0);
     }
 
+    if (expect_application_protocol_match && expect_source_type_test) {
+      if (absl::StartsWith(source_address, "/")) {
+        remote_address_.reset(new Network::Address::PipeInstance(source_address));
+      } else {
+        remote_address_.reset(new Network::Address::Ipv4Instance(source_address, 111));
+      }
+      EXPECT_CALL(*socket_, remoteAddress()).Times(1).WillRepeatedly(ReturnRef(remote_address_));
+    } else {
+      EXPECT_CALL(*socket_, remoteAddress()).Times(0);
+    }
+
     return manager_->listeners().back().get().filterChainManager().findFilterChain(*socket_);
```

Comment:
Could you add a guard here to make sure that all expectations matched, i.e.

```
if (expect_destination_port_match && expect_destination_address_match &&
    expect_server_name_match && expect_transport_protocol_match &&
    expect_application_protocol_match) {
  return manager_->listeners().back().get().filterChainManager().findFilterChain(*socket_);
} else {
  return nullptr;
}
```

This will catch the fact that you're passing `expect_application_protocol_match=false` as an input, when you're really expecting it to be `true` in all cases that result in a valid filter chain.

Thanks!

Reply:
@PiotrSikora I'm going to add this check in the beginning of `findFilterChain, otherwise we set function call expectations that are not satisfied. Like this:

```
Actual function call count doesn't match EXPECT_CALL(*socket_, requestedServerName())...
         Expected: to be called once
           Actual: never called - unsatisfied and active
```

Reply:
I'm an idiot, I completely missed the fact that it's doing the actual check... Sorry for sending you on the wrong path!

Something like that should work:
```
const Network::FilterChain* result = manager_->listeners().back().get().filterChainManager().findFilterChain(*socket_);
if (expect_destination_port_match && expect_destination_address_match &&
    expect_server_name_match && expect_transport_protocol_match &&
    expect_application_protocol_match) {
  EXPECT_NE(result, nullptr);
} else {
  EXPECT_EQ(result, nullptr);
}
return result;
```

Reply:
@PiotrSikora I had to add `expect_source_type_match` to allow semantics for expressing when the source type should match. This is different from `expect_source_type_test` since it describes when the source type will be checked, but does not necessarily mean match per se.